[Cases] Persist the cases table state on the URL

cnasikas opened the pull request:
## Summary

This PR persists the whole state of the cases table in the URL. Specifically:

- The format of the URL changed. The Rison format will be used which is widely used in Kibana.
- The old format will be supported to support BWC.
- All filters are persisted in the URL including custom fields.
- The URL is the source of truth and it takes precedence over everything.
- The state is also persisted in the local storage. The state from the local storage is loaded only when the URL is empty.
- The state in the local storage is stored under a new key called `cases.list.state`. The old key will not be taken into account.
- Navigate through pages and back to the cases table do not lose the state of the table.
- 

## Testing

- Test that the filtering is working as expected with all possible filters and values.
- Test that the URL is updating correctly as you change the filters and the pagination.
- Test that submitting a URL takes priority over the state persisted in the local storage.
- Test that submitting a URL changes the filters and the pagination as expected.
- Test that submitting a URL makes any hidden filters visible.
- Test that legacy URLs are working as expected.
- Put malformed values and keys in the URL and see how the application behaves.
- Test that the extra query params put by the Security solution persisted and do not affect the cases filtering.
- Test that the configuration of the filters (which ones are visible and which ones are hidden) is not affected.
- Hide all filters and put a URL that contains all filters. They should be shown.
- Remove the local storage state and check how the application behaves when you enter a URL with filters
- Ensure that when you navigate between pages your filtering is persisted.
- Ensure that the state from the local storage is loaded when you enter a URL without query parameters.
- No assignees filtering is working.
- Passing non-existing custom fields on the URL does not lead to an error.
- Paste a URL that is not the same as the state in the local storage, navigate to a case, and come back to the cases table. The filters set by the URL should be the same.

Blocked by: https://github.com/elastic/kibana/pull/176546
Flaky test runner: https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/5114, https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/5119

### Checklist

Delete any items that are not applicable to this PR.

- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [x] [Flaky Test Runner](https://ci-stats.kibana.dev/trigger_flaky_test_runner/1) was used on any tests changed

### For maintainers

- [x] This was checked for breaking API changes and was [labeled appropriately](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)

## Release notes

Persist all filter options of the cases table, including custom fields, in the URL. The filtering also persists when navigating back and forth between pages.

cnasikas commented on x-pack/plugins/cases/common/constants/index.ts:
```diff
@@ -201,10 +201,9 @@ export const SEARCH_DEBOUNCE_MS = 500;
-  casesFilterOptions: 'cases.list.filterOptions',
   casesTableColumns: 'cases.list.tableColumns',
   casesTableFiltersConfig: 'cases.list.tableFiltersConfig',
+  casesTableState: 'cases.list.state',
```
There will be one state in local storage that will contain both the query params and the filers options.

cnasikas commented on x-pack/plugins/cases/public/components/all_cases/all_cases_list.tsx:
```diff
@@ -215,6 +219,17 @@ export const AllCasesList = React.memo<AllCasesListProps>(
           currentUserProfile={currentUserProfile}
           filterOptions={filterOptions}
         />
+        <CasesTableUtilityBar
```
Lift up the component from `CasesTable` to `AllCasesList` to avoid prop drilling.

cnasikas commented on x-pack/plugins/cases/public/components/all_cases/use_all_cases_state.test.tsx:
```diff
@@ -74,225 +48,142 @@ describe('useAllCasesQueryParams', () => {
-    expect(result.current.filterOptions).toStrictEqual(DEFAULT_FILTER_OPTIONS);
-  });
-
-  it('updates localstorage with default values on first run', () => {
```
local storage is not updated on the first run. it is updated when the user interacts with the filters and the URL changes.

adcoelho replied:
This changed in the latest commit right? Is the test still relevant?

cnasikas replied:
Good question. We do not update the local storage with default values. We only update the local storage if the URL is not empty. The test that tests this behavior is called `updates the local storage when navigating to a URL and the query params are not empty`.

cnasikas commented:
/ci

cnasikas commented:
/ci

cnasikas commented:
/ci

cnasikas commented:
/ci

elasticmachine commented:
Pinging @elastic/response-ops (Team:ResponseOps)

adcoelho commented:
I found a small bug that isn't necessarily this PRs fault.

If I am filtering by a custom field and the custom field gets deleted, a toaster error will always pop up saying the custom field does not exist and the only way to fix things is to manually update the URL.

Maybe when parsing the fields in the URL we could match the keys against the ones in the configuration and not include them if they are missing?

cnasikas commented:
@adcoelho I fixed that in [`8de64e9` (#175237)](https://github.com/elastic/kibana/pull/175237/commits/8de64e974431be0dc852809d961c2c49ea5b710e). Could you please fetch the latest changes and test again?

adcoelho commented:
> @adcoelho I fixed that in [`8de64e9` (#175237)](https://github.com/elastic/kibana/pull/175237/commits/8de64e974431be0dc852809d961c2c49ea5b710e). Could you please fetch the latest changes and test again?


That fixed it but the field is still in the URL and the local storage 😅

adcoelho commented on x-pack/plugins/cases/public/components/all_cases/use_all_cases_state.tsx:
```diff
@@ -5,244 +5,183 @@
+
+  // use of useEffect because setUrlState calls history.push
+  useEffect(() => {
+    if (
+      !isStateLoadedFromLocalStorage.current &&
+      isURLStateEmpty(urlState) &&
+      localStorageState &&
+      !deepEqual(urlState, localStorageState) &&
+      !isModalView
+    ) {
```
One of the test scenarios is:

> Test that submitting a URL takes priority over the state persisted in the local storage.

This works as expected. 👍 

But defining a URL does not overwrite the local storage and I think it should.*

The local storage is only ever updated when a filter is clicked but what if I write a specific URL and go there? I would probably want it to be saved too.

In that scenario, the whole page is rendered but `set` is never called.

Maybe we could do:
```suggestion
    if (
      !isStateLoadedFromLocalStorage.current &&
      localStorageState &&
      !deepEqual(urlState, localStorageState) &&
      !isModalView
    ) {
```

* This ties in to [this comment](https://github.com/elastic/kibana/pull/175237#issuecomment-1934154403). If I have some "broken" field in the URL I can only remove it by deleting URL and local storage.

cnasikas replied:
Good point. Fixed in [`feae267` (#175237)](https://github.com/elastic/kibana/pull/175237/commits/feae26737b5a996c8229ebd64c6703050384108f).

cnasikas commented:
> > @adcoelho I fixed that in [`8de64e9` (#175237)](https://github.com/elastic/kibana/pull/175237/commits/8de64e974431be0dc852809d961c2c49ea5b710e). Could you please fetch the latest changes and test again?
> 
> That fixed it but the field is still in the URL and the local storage 😅

That's a different remark 🙂. It is by design like this. The moment you interact with the filters the URL and the local storage will change.

cnasikas commented on x-pack/plugins/cases/public/components/all_cases/schema.ts:
```diff
@@ -0,0 +1,46 @@
+    tags: rt.array(rt.string),
+    category: rt.array(rt.string),
+    assignees: rt.array(rt.union([rt.string, rt.null])),
+    customFields: rt.record(rt.string, rt.array(rt.string)),
```
A lot of the values will be dynamic (configured by users). Think of the list custom field type for example. It happened for the toggle to be static.

adcoelho commented:
Tested everything and mostly works as expected. One bug was fixed already. I still wanna take another look at the code.

There is only one issue I could find with the search box(in bold below).

- [x]  Test that the filtering is working as expected with all possible filters and values.
- [x]  Test that the URL is updating correctly as you change the filters and the pagination.
- [x]  Test that submitting a URL takes priority over the state persisted in the local storage.
    - Yes but doesn’t overwrite it. Fixed in `[feae267` (#175237)](https://github.com/elastic/kibana/pull/175237/commits/feae26737b5a996c8229ebd64c6703050384108f).
- [x]  Test that submitting a URL changes the filters and the pagination as expected.
    - **Defining a search term does not populate the search box if I navigate out and back.**
- [x]  Test that submitting a URL makes any hidden filters visible.
- [x]  Test that legacy URLs are working as expected.
- [x]  Put malformed values and keys in the URL and see how the application behaves.
- [x]  Test that the extra query params put by the Security solution persisted and do not affect the cases filtering.
- [x]  Test that the configuration of the filters (which ones are visible and which ones are hidden) is not affected.
- [x]  Hide all filters and put a URL that contains all filters. They should be shown.
- [x]  Remove the local storage state and check how the application behaves when you enter a URL with filters
    - Same as above. Fixed in `[feae267` (#175237)](https://github.com/elastic/kibana/pull/175237/commits/feae26737b5a996c8229ebd64c6703050384108f).
- [x]  Ensure that when you navigate between pages your filtering is persisted.
- [x]  Ensure that the state from the local storage is loaded when you enter a URL without query parameters.
- [x]  No assignees filtering is working.
- [x]  Passing non-existing custom fields on the URL does not lead to an error.

adcoelho commented on x-pack/plugins/cases/public/components/all_cases/schema.ts:
```diff
@@ -0,0 +1,46 @@
+      rt.literal('title'),
+      rt.literal('category'),
+    ]),
+    page: rt.number,
+    perPage: rt.number,
```
Maybe nit but when playing with the URL I thought we could limit these two.

At least there is no point in allowing page to be `<= 2`.

As for the `perPage` the UI limits it to 10, 25 etc so we could do the same 🤷

adcoelho commented on x-pack/plugins/cases/public/components/all_cases/table_filters.test.tsx:
```diff
@@ -954,35 +963,111 @@ describe('CasesTableFilters ', () => {
+        const button = await screen.findByRole('button', { name: filter.name });
+        expect(button).toBeInTheDocument();
+        expect(
+          within(button).getByLabelText(`${filter.active} active filters`)
```
```suggestion
          within(button).findByLabelText(`${filter.active} active filters`)
```

adcoelho commented on x-pack/plugins/cases/public/components/all_cases/table_filters.test.tsx:
```diff
@@ -954,35 +963,111 @@ describe('CasesTableFilters ', () => {
+        const button = await screen.findByRole('button', { name: filter.name });
+        expect(button).toBeInTheDocument();
+        expect(
+          within(button).getByLabelText(`${filter.active} active filters`)
```
```suggestion
          within(button).findByLabelText(`${filter.active} active filters`)
```

adcoelho commented on x-pack/plugins/cases/public/components/all_cases/table_filters.test.tsx:
```diff
@@ -954,35 +963,111 @@ describe('CasesTableFilters ', () => {
+        },
+      };
+
+      appMockRender = createAppMockRenderer({ license });
```
Why creating it again?

adcoelho commented on x-pack/plugins/cases/public/components/all_cases/table_filters.test.tsx:
```diff
@@ -378,7 +375,7 @@ describe('CasesTableFilters ', () => {
       expect(screen.getByTestId('options-filter-popover-button-assignees')).toBeInTheDocument();
```
In the continuing effort to change the files we have to touch maybe these could also be changed? 😅 

```suggestion
      expect(screen.findByTestId('options-filter-popover-button-assignees')).toBeInTheDocument();
```

adcoelho commented on x-pack/plugins/cases/public/components/all_cases/table_filters.test.tsx:
```diff
@@ -764,6 +766,7 @@ describe('CasesTableFilters ', () => {
         { key: 'severity', isActive: true },
       ];
+
       localStorage.setItem(
```
I am a bit(very) confused but how is the local storage relevant here? 

My understanding is it only affects the values in `useAllCasesState` and that is done **above** table filters. That url/localstorage "state" is then passed to the component as props.

cnasikas replied:
The local storage here is used to test the filter configuration (which filters are active and which are not) not the state. The `CasesTableFilters` uses the `useFilterConfig` which in turn uses the local storage to load the active filters.

adcoelho commented on x-pack/plugins/cases/public/components/all_cases/utils/all_cases_url_state_deserializer.test.ts:
```diff
@@ -0,0 +1,236 @@
+  });
+
+  it('sets the defaults to page and perPage correctly if they are not numbers', () => {
+    const url = 'page=foo&perPage=bar';
```
super nit: if we moved to Rison maybe that should become the default in tests too

cnasikas replied:
`allCasesUrlStateDeserializer` accepts an object so I used that as an input. The URLs are tested in `parseUrlParams`.

adcoelho reviewed (approved):
Tested and it works fine. Reviewed the code and left some comments.

However, the main comment I have is, please split these changes next time 😅 

All the utils could have been reviewed independently e.g.

js-jankisalvi reviewed (commented):
Wow.. Awesome work!! 🎉 
I am still going through code, tested it locally.

Verified below scenarios all worked fine except last one: 

- Test that the filtering is working as expected with all possible filters and values. ✅ 
- Test that the URL is updating correctly as you change the filters and the pagination. ✅ 
- Test that submitting a URL takes priority over the state persisted in the local storage. ✅ 
- Test that submitting a URL changes the filters and the pagination as expected. ✅ 
- Test that submitting a URL makes any hidden filters visible. ✅ 
- Test that legacy URLs are working as expected. ✅ 
- Put malformed values and keys in the URL and see how the application behaves. ✅ 
- Test that the extra query params put by the Security solution persisted and do not affect the cases filtering. ✅ 
- Test that the configuration of the filters (which ones are visible and which ones are hidden) is not affected. ✅ 
- Hide all filters and put a URL that contains all filters. They should be shown. ✅ 
- Remove the local storage state and check how the application behaves when you enter a URL with filters ✅ 
- Ensure that when you navigate between pages your filtering is persisted. ✅ 
- Ensure that the state from the local storage is loaded when you enter a URL without query parameters. ✅ 
- No assignees filtering is working. ✅ 
- Passing non-existing custom fields on the URL does not lead to an error. ✅ 
- Paste a URL that is not the same as the state in the local storage, navigate to a case, and come back to the cases table. The filters set by the URL should be the same. ❌ 

For the last point I found a bug regarding `customField`, 
- paste a URL with customField filter
- navigate to case and back 
- the customField filter is not there anymore:

https://github.com/elastic/kibana/assets/117571355/af4e284d-4a26-416c-afa7-3b26f8e7679a

cnasikas commented:
@adcoelho 

> However, the main comment I have is, please split these changes next time 😅

You are right, sorry for that!

> Defining a search term does not populate the search box if I navigate out and back.

Fixed in [`d9ea9a1` (#175237)](https://github.com/elastic/kibana/pull/175237/commits/d9ea9a1789d918d96162dd4b7f74ebe50cd114de)

@js-jankisalvi 

> For the last point I found a bug regarding customField,

Fixed in [`16b0812` (#175237)](https://github.com/elastic/kibana/pull/175237/commits/16b081220ad6ce5b13c3f8555ac35c82d08cabfa)

js-jankisalvi reviewed (commented):
Code looks amazing!! 👍 
However regarding the bug of customField, I think it is not completely fixed. I verified it after fetching latest changes.

It works fine if you are paste a URL and then refresh (so that localStorage loads customField), but if you paste to a fresh new tab and navigate to case and bak it still fails to persist customField value.
With refresh: 

https://github.com/elastic/kibana/assets/117571355/d7db6ce9-7766-4b1e-9149-ecdfa6f2bfd2

Without refresh:

https://github.com/elastic/kibana/assets/117571355/bb5aac43-4504-43f4-b944-c04b9ab5b738

js-jankisalvi reviewed (approved):
> However regarding the bug of customField, I think it is not completely fixed. I verified it after fetching latest changes.

As discussed offline, this bug can be taken care separately.

Great work!!

cnasikas commented:
Thanks @js-jankisalvi! I fixed it in [`86211c2` (#175237)](https://github.com/elastic/kibana/pull/175237/commits/86211c2bedb927db2393c75fb5ec91883330159d) 🙂
